domain/robstoll-lib/atrium-domain-robstoll-lib-jvm/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/lib/creating/filesystem/symbolicLinkResolving.kt

```diff
@@ -0,0 +1,101 @@
+import ch.tutteli.atrium.translations.DescriptionPathAssertion.FAILURE_DUE_TO_LINK_LOOP
+import ch.tutteli.atrium.translations.DescriptionPathAssertion.HINT_FOLLOWED_SYMBOLIC_LINK
+import java.io.IOException
+import java.nio.file.Files
+import java.nio.file.Path
+import java.util.*
+
+inline fun explainForResolvedLink(path: Path, block: (realPath: Path) -> Assertion): Assertion {
+    val hintList = LinkedList<Assertion>()
+    val realPath = addAllLevelResolvedSymlinkHints(
+        path.toAbsolutePath(),
+        hintList
+    )
+    val blockResult = block(realPath)
+    return if (hintList.isNotEmpty()) {
+        when (blockResult) {
+            is AssertionGroup -> hintList.addAll(blockResult.assertions)
+            else -> hintList.add(blockResult)
+        }
+        ExpectImpl.builder.explanatoryGroup.withDefaultType
+            .withAssertions(hintList)
+            .build()
+    } else {
+        blockResult
+    }
+}
+
+/**
+ * Resolves the provided [absolutePath] and returns the resolved target (if resolving is possible).
+ * Adds explanatory hints for all involved symbolic links to [hintList].
+ */
+fun addAllLevelResolvedSymlinkHints(
+    absolutePath: Path,
+    hintList: Deque<Assertion>
+) = addAllLevelResolvedSymlinkHints(absolutePath, hintList, LinkedList(), absolutePath)
+
+private fun addAllLevelResolvedSymlinkHints(
+    absolutePath: Path,
+    hintList: Deque<Assertion>,
+    visitedList: MutableList<Path>,
+    initialPath: Path
+): Path {
+    var currentPath = absolutePath.root
+    visitedList.add(absolutePath)
+
+    for (part in absolutePath) {
+        currentPath = currentPath.resolve(part)
+        val nextPath = addOneStepResolvedSymlinkHint(
+            currentPath,
+            hintList
+        )
+        if (nextPath != null) {
+            val visitedIndex = visitedList.indexOf(nextPath)
+            if (visitedIndex != -1) {
+                repeat(hintList.size - visitedIndex) { hintList.removeLast() }
+                // add to the list so [hintForLinkLoop] prints this duplicate twice
+                visitedList.add(nextPath)
+                hintList.add(hintForLinkLoop(visitedList, visitedIndex))
+                return initialPath
+            } else {
+                currentPath = addAllLevelResolvedSymlinkHints(nextPath, hintList, visitedList, initialPath)
+            }
+        }
+    }
+    return currentPath
+}
+
+/**
+ * If [absolutePath] is surely a symlink, adds an explanatory hint to [resultList] and returns the link target.
+ * Return `null` and does not modify [resultList] otherwise.
+ */
+private fun addOneStepResolvedSymlinkHint(absolutePath: Path, resultList: Deque<Assertion>): Path? {
+    try {
+        val nextPath = Files.readSymbolicLink(absolutePath)
+        resultList.add(
+            ExpectImpl.builder.explanatory
+                .withExplanation(HINT_FOLLOWED_SYMBOLIC_LINK, absolutePath, nextPath)
+                .build()
+        )
+        return nextPath
```

Comment:
I would remove both `return` and turn this method into a method with expression body. But up to you

Reply:
Done.